[Serve][Doc] Add custom request router docs

GeneDer opened the pull request:
<!-- Thank you for your contribution! Please review https://github.com/ray-project/ray/blob/master/CONTRIBUTING.rst before opening a pull request. -->

<!-- Please add a reviewer to the assignee section when you create a PR. If you don't have the access to it, we will shortly find a reviewer and assign them to your PR. -->

## Why are these changes needed?

Add example docs around using custom request router.

## Related issue number

Closes https://anyscale1.atlassian.net/browse/LLM-1978?atlOrigin=eyJpIjoiZmVmMzQzNzlkY2RlNGMzZGJhYWVhMDMwYjI3NDY3NjIiLCJwIjoiaiJ9

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a
           method in Tune, I've added it in `doc/source/tune/api/` under the
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

GeneDer committed:
WIP

Signed-off-by: Gene Su <e870252314@gmail.com>

GeneDer committed:
Merge branch 'master' into prototype-custom-replica-scheduler

GeneDer committed:
WIP: refactor locality minxin

Signed-off-by: Gene Su <e870252314@gmail.com>

GeneDer committed:
use context vars for tracking flags

Signed-off-by: Gene Su <e870252314@gmail.com>

GeneDer committed:
Merge branch 'master' into prototype-custom-replica-scheduler

GeneDer committed:
serialize replica scheduler

Signed-off-by: Gene Su <e870252314@gmail.com>

GeneDer committed:
fix update mulitplexed mode id

Signed-off-by: Gene Su <e870252314@gmail.com>

GeneDer committed:
more refactor

Signed-off-by: Gene Su <e870252314@gmail.com>

GeneDer committed:
WIP: still breaking, need to test passing the _RequestSchedulingContext directly

Signed-off-by: Gene Su <e870252314@gmail.com>

GeneDer committed:
Merge branch 'master' into prototype-custom-replica-scheduler

GeneDer committed:
Merge branch 'master' into prototype-custom-replica-scheduler

GeneDer committed:
add get_deployment_config api and use the replica scheduler set onto the deployment config

Signed-off-by: Gene Su <e870252314@gmail.com>

GeneDer committed:
lint

Signed-off-by: Gene Su <e870252314@gmail.com>

GeneDer committed:
fix getting deployment config

Signed-off-by: Gene Su <e870252314@gmail.com>

GeneDer committed:
fix the key passed into choose_replicas

Signed-off-by: Gene Su <e870252314@gmail.com>

GeneDer committed:
WIP

Signed-off-by: Gene Su <e870252314@gmail.com>

GeneDer committed:
WIP: fix some tests

Signed-off-by: Gene Su <e870252314@gmail.com>

GeneDer committed:
fix all tests

Signed-off-by: Gene Su <e870252314@gmail.com>

GeneDer committed:
fix import

Signed-off-by: Gene Su <e870252314@gmail.com>

GeneDer committed:
WIP

Signed-off-by: Gene Su <e870252314@gmail.com>

GeneDer committed:
refactor common scheduler logics

Signed-off-by: Gene Su <e870252314@gmail.com>

GeneDer committed:
fix

Signed-off-by: Gene Su <e870252314@gmail.com>

GeneDer committed:
enable user to pass scheduling stats

Signed-off-by: Gene Su <e870252314@gmail.com>

GeneDer committed:
try again

Signed-off-by: Gene Su <e870252314@gmail.com>

GeneDer committed:
create alias for public replica scheduler module

Signed-off-by: Gene Su <e870252314@gmail.com>

GeneDer committed:
fix choose_replicas interface and add scheduling_stats property

Signed-off-by: Gene Su <e870252314@gmail.com>

GeneDer committed:
fix type

Signed-off-by: Gene Su <e870252314@gmail.com>

GeneDer committed:
expose ReplicaID and ReplicaSchedulingInfo publically

Signed-off-by: Gene Su <e870252314@gmail.com>

GeneDer committed:
expose the replica scheduler fields publically and allow the deployment decroator to set it

Signed-off-by: Gene Su <e870252314@gmail.com>

GeneDer committed:
Merge branch 'master' into prototype-custom-replica-scheduler

GeneDer committed:
allow to reconfigure replica scheduler for a given handle

Signed-off-by: Gene Su <e870252314@gmail.com>

GeneDer committed:
add on_request_scheduled callback

Signed-off-by: Gene Su <e870252314@gmail.com>

GeneDer committed:
allow the replica scheduler to configure to match on the exact request

Signed-off-by: Gene Su <e870252314@gmail.com>

GeneDer committed:
refactor choose_replicas into choose_replicas_with_backoff and choose_replicas

Signed-off-by: Gene Su <e870252314@gmail.com>

GeneDer committed:
implement rank_replicas on multiplexed and locality mixin and update choose_replicas interface

Signed-off-by: Gene Su <e870252314@gmail.com>

GeneDer committed:
create FIFOMixin and change the default to OOO scheduling

Signed-off-by: Gene Su <e870252314@gmail.com>

GeneDer committed:
add select_available_replicas to filter replicas unavailable to take more requests

Signed-off-by: Gene Su <e870252314@gmail.com>

GeneDer committed:
uncomment commented code

Signed-off-by: Gene Su <e870252314@gmail.com>

GeneDer committed:
fix assign_request on LocalRouter

Signed-off-by: Gene Su <e870252314@gmail.com>

GeneDer committed:
Merge branch 'master' into prototype-custom-replica-scheduler

GeneDer committed:
call on_request_scheduled directly instead of allowing setting with .options

Signed-off-by: Gene Su <e870252314@gmail.com>

GeneDer committed:
fix on_request_scheduled interface

Signed-off-by: Gene Su <e870252314@gmail.com>

GeneDer committed:
Merge branch 'master' into prototype-custom-replica-scheduler

GeneDer committed:
allow the controller to update scheduling stats

Signed-off-by: Gene Su <e870252314@gmail.com>

GeneDer committed:
fixes

Signed-off-by: Gene Su <e870252314@gmail.com>

GeneDer committed:
update the choose_replicas to be a list of list instead of list of set to preserve ordering

Signed-off-by: Gene Su <e870252314@gmail.com>

GeneDer committed:
add requestSchedulingStatsPeriodS and requestSchedulingStatsTimeoutS on java

Signed-off-by: Gene Su <e870252314@gmail.com>

GeneDer committed:
fix controller test

Signed-off-by: Gene Su <e870252314@gmail.com>

GeneDer committed:
implement get_scheduling_stats onto MockReplicaActorWrapper

Signed-off-by: Gene Su <e870252314@gmail.com>

GeneDer committed:
update interface for ranking helpers

Signed-off-by: Gene Su <e870252314@gmail.com>

GeneDer committed:
lint

Signed-off-by: Gene Su <e870252314@gmail.com>

GeneDer committed:
fix recording scheduling stats

Signed-off-by: Gene Su <e870252314@gmail.com>

GeneDer committed:
expoxe ReplicaResult publically

Signed-off-by: Gene Su <e870252314@gmail.com>

GeneDer committed:
Merge branch 'master' into prototype-custom-replica-scheduler

GeneDer committed:
[Serve] Implement recording routing stats periodically

Signed-off-by: Gene Su <e870252314@gmail.com>

GeneDer committed:
checkout ci/lint/pydoclint-baseline.txt

Signed-off-by: Gene Su <e870252314@gmail.com>

GeneDer committed:
add missing protobuf

Signed-off-by: Gene Su <e870252314@gmail.com>

GeneDer committed:
add missing const

Signed-off-by: Gene Su <e870252314@gmail.com>

GeneDer committed:
add more missing changes

Signed-off-by: Gene Su <e870252314@gmail.com>

GeneDer committed:
add more missing changes

Signed-off-by: Gene Su <e870252314@gmail.com>

GeneDer committed:
lint

Signed-off-by: Gene Su <e870252314@gmail.com>

GeneDer committed:
Merge branch 'master' into implement-routing-stats

GeneDer committed:
WIP: add test and fix missing changes

Signed-off-by: Gene Su <e870252314@gmail.com>

GeneDer committed:
fix tests

Signed-off-by: Gene Su <e870252314@gmail.com>

GeneDer committed:
ensure test_record_routing_stats.py is ran

Signed-off-by: Gene Su <e870252314@gmail.com>

GeneDer committed:
fix test

Signed-off-by: Gene Su <e870252314@gmail.com>

GeneDer committed:
Merge branch 'master' into prototype-custom-replica-scheduler

GeneDer committed:
revert unnecessary changes and add ranking helpers

Signed-off-by: Gene Su <e870252314@gmail.com>

GeneDer committed:
Merge branch 'master' into prototype-custom-replica-scheduler

GeneDer committed:
Merge branch 'master' into prototype-custom-replica-scheduler

GeneDer committed:
add tests for the ranking helpers

Signed-off-by: Gene Su <e870252314@gmail.com>

GeneDer committed:
Merge branch 'prototype-custom-replica-scheduler' into add-custom-request-router-docs

GeneDer committed:
Merge branch 'master' into add-custom-request-router-docs

GeneDer committed:
[Serve] add example docs to use custom request router

Signed-off-by: Gene Su <e870252314@gmail.com>

GeneDer committed:
fix

Signed-off-by: Gene Su <e870252314@gmail.com>

GeneDer committed:
try again

Signed-off-by: Gene Su <e870252314@gmail.com>

GeneDer committed:
try again

Signed-off-by: Gene Su <e870252314@gmail.com>

GeneDer committed:
complete the docs

Signed-off-by: Gene Su <e870252314@gmail.com>

GeneDer committed:
use 0 cpu resources

Signed-off-by: Gene Su <e870252314@gmail.com>

GeneDer committed:
fix return types

Signed-off-by: Gene Su <e870252314@gmail.com>

GeneDer committed:
test not to call .result()

Signed-off-by: Gene Su <e870252314@gmail.com>

GeneDer committed:
Merge branch 'master' into add-custom-request-router-docs

GeneDer committed:
test to run the event in the asyncio router loop

Signed-off-by: Gene Su <e870252314@gmail.com>

GeneDer committed:
test comment out request_router_class

Signed-off-by: Gene Su <e870252314@gmail.com>

GeneDer committed:
test splitting out request router definition and app definition

Signed-off-by: Gene Su <e870252314@gmail.com>

GeneDer committed:
update UniformRequestRouter code and revert event loop is running check

Signed-off-by: Gene Su <e870252314@gmail.com>

GeneDer committed:
address comments

Signed-off-by: Gene Su <e870252314@gmail.com>

GeneDer committed:
add docs around mixins

Signed-off-by: Gene Su <e870252314@gmail.com>

GeneDer committed:
address comments

Signed-off-by: Gene Su <e870252314@gmail.com>

GeneDer committed:
Merge branch 'master' into add-custom-request-router-docs

GeneDer committed:
Merge branch 'master' into add-custom-request-router-docs

GeneDer committed:
cleanup

Signed-off-by: Gene Su <e870252314@gmail.com>

GeneDer committed:
add request router to Serve's api doc

Signed-off-by: Gene Su <e870252314@gmail.com>

GeneDer committed:
fix typo

Signed-off-by: Gene Su <e870252314@gmail.com>

GeneDer committed:
fix link

Signed-off-by: Gene Su <e870252314@gmail.com>

GeneDer committed:
Merge branch 'master' into add-custom-request-router-docs

GeneDer committed:
add docs for dataclasses and update links in the doc

Signed-off-by: Gene Su <e870252314@gmail.com>

GeneDer committed:
fix MultiplexMixin naming

Signed-off-by: Gene Su <e870252314@gmail.com>

GeneDer committed:
fix rank_replicas_via_locality link

Signed-off-by: Gene Su <e870252314@gmail.com>

GeneDer commented:
@abrarsheikh can you help to do a first pass on this one too plz. cc @kouroshHakha hopefully all the links should work with this latest commit

kouroshHakha commented:
We should be very deliberate about public vs. private methods on the API side. Do not over publish something that we are not sure should remain private or public. (make it private for now)

![image](https://github.com/user-attachments/assets/c713b3d6-717c-4487-856a-64c72ea20ff7)

kouroshHakha commented:
My comment applies to all the items in this list: 

![image](https://github.com/user-attachments/assets/df148383-043f-4eaa-a2bb-5b63dae0b96f)

kouroshHakha commented:
Links in this section are not linked: 
![image](https://github.com/user-attachments/assets/2b779f08-0300-4c08-9c47-76b4170cce55)

GeneDer committed:
add links tp the mixin section

Signed-off-by: Gene Su <e870252314@gmail.com>

GeneDer committed:
make some methods private

Signed-off-by: Gene Su <e870252314@gmail.com>

GeneDer committed:
make more method private

Signed-off-by: Gene Su <e870252314@gmail.com>

GeneDer commented on doc/source/serve/doc_code/custom_request_router.py:
```diff
@@ -0,0 +1,123 @@
+# flake8: noqa
```
I think this is due to we have multiple code blocks in the same file, and flake8 will try to complain those. I'm pretty sure there are still come formatter running through those code as you can find it in some of the earlier failures https://buildkite.com/ray-project/microcheck/builds/17674#019733f9-acf4-412e-83a8-817c01311d62/187-435

Also those code are tested like https://buildkite.com/ray-project/premerge/builds/41373#01973bbc-a562-4fc8-94f2-8cb3fc9e2de5/190-1124

GeneDer commented:
> @GeneDer @kouroshHakha Correct me if I'm wrong, but in the current state of this PR, ALL the methods in the RequestRouter are marked as private, so all links back to the RequestRouter API docs won't resolve.
> 
> Please advise; I've started a round of robust edits on the docs, but these don't appear to be viable. I am abandoning this work until it's clear what is actually public. Can we target a docs release next week after we schedule time to discuss what's actually being shipped?

No, I don't think all the methods should be marked private, only the ones that's not intended to be used by the users should mark private. I already addressed those in my latest commit so no need to revisit this again. cc @kouroshHakha please take another look when you have some time

GeneDer committed:
try to add doc for all attributes

Signed-off-by: Gene Su <e870252314@gmail.com>

GeneDer committed:
mark request router related components alpha api

Signed-off-by: Gene Su <e870252314@gmail.com>

kouroshHakha commented:
@GeneDer from my side, I have two concerns:

1. Please mark this apis as public (on both the guide and the serve api section)
2. Please file follow up doc tickets that we should do the following (please be more specific and concrete than I am here :))
- We only need to expose what the user would need to override for 90% of usecases. I don't want to document the methods that already have a deep implementation which we think users will never override. There is a distincition between public methods (from caller's perspective) and what needs to be overriden by user. In my opinion RequestRouter is a private class that can be overriden which is not a well defined concept in ray docs hence the confusion. 
<img width="1148" alt="image" src="https://github.com/user-attachments/assets/9f30ee1a-736d-4d60-b4b5-51117e3af049" />

- Due to the same reasoning as above, we need to clean up the attributes as well, and only document those that a user would need during overriding the class.
<img width="1127" alt="image" src="https://github.com/user-attachments/assets/7d3516be-02e3-465e-904c-7b2cc643930a" />

- From these subset of apis, again, only those methods / attributes that user would need in order to complete RequestRouting customization would need to be documented.

<img width="1135" alt="image" src="https://github.com/user-attachments/assets/333897dd-402a-462e-9300-8423cbc7a8d4" />

- Having no method in FifoMixin is confusing. Then what is this mixin? How does it work?

<img width="1167" alt="image" src="https://github.com/user-attachments/assets/c7e5a295-4b9b-4bba-937f-1c147bb8d799" />

GeneDer committed:
add alpha warning to the guide and add more descriptions on FIFO mixin

Signed-off-by: Gene Su <e870252314@gmail.com>

GeneDer commented:
@kouroshHakha addressed all your comments and created issues at https://anyscale1.atlassian.net/browse/LLM-1991, https://anyscale1.atlassian.net/browse/LLM-1992, and https://anyscale1.atlassian.net/browse/LLM-1993. For FIFO mixin I just update the docstring to describe it better hopefully no need follow up on that

GeneDer committed:
Merge branch 'master' into add-custom-request-router-docs

GeneDer committed:
try again

Signed-off-by: Gene Su <e870252314@gmail.com>

GeneDer committed:
add private classes to white list

Signed-off-by: Gene Su <e870252314@gmail.com>

GeneDer committed:
add ReplicaID to the white list

Signed-off-by: Gene Su <e870252314@gmail.com>

akshay-anyscale commented on doc/source/serve/advanced-guides/custom-request-router.md:
```diff
@@ -0,0 +1,151 @@
+implements the [`choose_replicas`](../api/doc/ray.serve.request_router.RequestRouter.choose_replicas.rst)
+method, which returns the random replica for all incoming requests. The returned type
+is a list of lists of replicas, where each inner list represents a rank of replicas.
+The first rank is the most preferred and the last rank is the least preferred.
```
We should explain this API a bit more - what happens upstream of this? Do we try until this list is exhausted, etc.?

akshay-anyscale commented on doc/source/serve/advanced-guides/custom-request-router.md:
```diff
@@ -0,0 +1,151 @@
+request router. These mixins can be used to implement common routing policies such as
+locality-aware routing, multiplexed model support, and FIFO request routing.
+
+- [`FIFOMixin`](../api/doc/ray.serve.request_router.FIFOMixin.rst): This mixin implements first in first out (FIFO)
```
Is this supposed to be a mixin or just a flag?

GeneDer replied:
This is a mixin that would alter the algorithm to do FIFO when included in the custom request router class. It's more then just a flag.

The default behavior is out of order matching the exact request in the queue with the one that's passed into `choose_replicas`.  Let me be even more explicit on it.

GeneDer committed:
add more details on the choose_replicas return value and FIFOMixin behavior

Signed-off-by: Gene Su <e870252314@gmail.com>
